src/ray/raylet/scheduling/cluster_task_manager_test.cc

```diff
@@ -398,6 +407,62 @@ TEST_F(ClusterTaskManagerTest, TestSpillAfterAssigned) {
   AssertNoLeaks();
 }
 
+TEST_F(ClusterTaskManagerTest, TestMinimizeWorkerPops) {
+  /*
+    This test ensures that we don't pop workers until the last second.
+    See https://github.com/ray-project/ray/issues/13725.
+  */
+  std::shared_ptr<MockWorker> worker =
+      std::make_shared<MockWorker>(WorkerID::FromRandom(), 1234);
+  pool_.PushWorker(worker);
+
+  int num_callbacks = 0;
+  auto callback = [&](Status, std::function<void()>, std::function<void()>) {
+    num_callbacks++;
+  };
+
+  /* Blocked on starting a worker. */
+  auto task = CreateTask({{ray::kCPU_ResourceLabel, 5}});
+  auto task2 = CreateTask({{ray::kCPU_ResourceLabel, 5}});
+
+  rpc::RequestWorkerLeaseReply local_reply;
+  task_manager_.QueueAndScheduleTask(task2, &local_reply, callback);
+  task_manager_.QueueAndScheduleTask(task, &local_reply, callback);
+
+  ASSERT_EQ(num_callbacks, 1);
+  ASSERT_EQ(leased_workers_.size(), 1);
+  ASSERT_EQ(pool_.num_pops, 1);
+
+  {
```

Comment:
Can we please use only public methods to modify the task manager's internal state?

Reply:
Yeah, just realized all this fancy dependency manager stuff is here and you already wrote the test I was going to write, so I added some asserts to your test instead.